### PR TITLE
Add variant of trufflehog that scans git history

### PR DESCRIPTION
**Pull request by det**
- Runs once on the repo root instead of individual files.
- Tells trufflehog what the upstream is.
- Trufflehog scans all commits in between.
- Ensure that timestamps are deterministic so that commit hashes are deterministic.

### CODE REVIEW

**Review by TylerJang27 — COMMENTED** [review]
Could we get some quick test coverage of this? Probably something similar to the `git-diff-check` test setup?

**Comment by det** [comment]
> Could we get some quick test coverage of this? Probably something similar to the `git-diff-check` test setup?

Thanks for the pointer, done.

File: linters/trufflehog/trufflehog.test.ts
Comment (TylerJang27) [inline linters/trufflehog/trufflehog.test.ts:11]: At least the first call shouldn't be necessary? Everything in the test_data folder gets copied into the sandbox

**Review by TylerJang27 — APPROVED** [review]
LGTM. If the commit SHA still isn't playing nice in the tests, you can modify the `normalizeMessage` call to look for `commit <sha>` and replace `<sha>`, although we might lose a little bit of coverage

File: linters/trufflehog/trufflehog.test.ts
Comment (det) [reply linters/trufflehog/trufflehog.test.ts:11]: I think `customLinterCheckTest` doesn't copy those files? The `git add` fails without copying it myself first.

**Comment by det** [comment]
> LGTM. If the commit SHA still isn't playing nice in the tests, you can modify the `normalizeMessage` call to look for `commit <sha>` and replace `<sha>`, although we might lose a little bit of coverage

Determinism was more effort than it was worth. Added the ability to normalize the landing state after running linters to `customLinterCheckTest`. I'll wait for your sign off before landing.

File: linters/trufflehog/trufflehog.test.ts
Comment (TylerJang27) [reply linters/trufflehog/trufflehog.test.ts:12]:
```
    await driver.setUp();
    if (preCheck) {
      // preCheck is not always async, but we must await in case it is.
      await preCheck(driver);
      driver.debug("Finished running custom preCheck hook");
    }
```
preCheck should be called after any initial linter driver setup, so it should be there? Potentially worth investigating but priority is getting this landed

File: linters/trufflehog/trufflehog.test.ts
Comment (TylerJang27) [inline linters/trufflehog/trufflehog.test.ts:25]: Mild preference for this in `landing_state.ts` rather than as an additional input to `customLinterCheckTest`, but this should be fine and not a strong opinion.

**Review by TylerJang27 — APPROVED** [review]

**Pull request merged**
